python/cugraph-pyg/cugraph_pyg/data/graph_store.py

```diff
@@ -0,0 +1,322 @@
+                world_size = torch.distributed.get_world_size()
+
+                vertices_array = cupy.arange(
+                    sum(self._num_vertices().values()), dtype="int64"
+                )
+                vertices_array = cupy.array_split(vertices_array, world_size)[rank]
+
+                self.__graph = pylibcugraph.MGGraph(
+                    self._resource_handle,
+                    graph_properties,
+                    [cupy.asarray(edgelist_dict["src"]).astype("int64")],
+                    [cupy.asarray(edgelist_dict["dst"]).astype("int64")],
+                    vertices_array=[vertices_array],
+                    edge_id_array=[cupy.asarray(edgelist_dict["eid"])],
+                    edge_type_array=[cupy.asarray(edgelist_dict["etp"])],
+                )
+            else:
+                self.__graph = pylibcugraph.SGGraph(
+                    self._resource_handle,
+                    graph_properties,
+                    cupy.asarray(edgelist_dict["src"]).astype("int64"),
+                    cupy.asarray(edgelist_dict["dst"]).astype("int64"),
+                    vertices_array=cupy.arange(
+                        sum(self._num_vertices().values()), dtype="int64"
+                    ),
+                    edge_id_array=cupy.asarray(edgelist_dict["eid"]),
+                    edge_type_array=cupy.asarray(edgelist_dict["etp"]),
+                )
+
+        return self.__graph
+
+    def _num_vertices(self) -> Dict[str, int]:
+        num_vertices = {}
+        for edge_attr in self.get_all_edge_attrs():
+            if edge_attr.size is not None:
+                num_vertices[edge_attr.edge_type[0]] = (
+                    max(num_vertices[edge_attr.edge_type[0]], edge_attr.size[0])
+                    if edge_attr.edge_type[0] in num_vertices
+                    else edge_attr.size[0]
+                )
+                num_vertices[edge_attr.edge_type[2]] = (
+                    max(num_vertices[edge_attr.edge_type[2]], edge_attr.size[1])
+                    if edge_attr.edge_type[2] in num_vertices
+                    else edge_attr.size[1]
+                )
+            else:
+                if edge_attr.edge_type[0] not in num_vertices:
+                    num_vertices[edge_attr.edge_type[0]] = int(
+                        self.__edge_indices[edge_attr.edge_type][0].max() + 1
+                    )
+                if edge_attr.edge_type[2] not in num_vertices:
+                    num_vertices[edge_attr.edge_type[1]] = int(
+                        self.__edge_indices[edge_attr.edge_type][1].max() + 1
+                    )
+
+        if self.is_multi_gpu:
+            vtypes = num_vertices.keys()
+            for vtype in vtypes:
+                sz = torch.tensor(num_vertices[vtype], device="cuda")
+                torch.distributed.all_reduce(sz, op=torch.distributed.ReduceOp.MAX)
+                num_vertices[vtype] = int(sz)
+        return num_vertices
```

Comment:
Should we cache `num_vertices` and `num_edges` since they require collective communications?

Reply:
We cache the offsets, which are more frequently used, but I might add caching the vertex and edge counts if they are used more often.  So maybe in the future.